File: src/Sample/Plumix.Sample/Demos/Material/CardDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/card_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class CardDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new CardDemoPageState();
    }
}

internal sealed class CardDemoPageState : State
{
    private bool _useMaterial3 = true;
    private bool _useThemeOverrides;
    private bool _clip;
    private bool _dense;

    public override Widget Build(BuildContext context)
    {
        var baseTheme = Theme.Of(context);
        var pageTheme = baseTheme with
        {
            UseMaterial3 = _useMaterial3,
            CardTheme = _useThemeOverrides
                ? new CardThemeData(
                    Color: Color.Parse("#FFF5F9EE"),
                    ShadowColor: Color.Parse("#FF455A64"),
                    SurfaceTintColor: Color.Parse("#FF6750A4"),
                    Elevation: 3,
                    Margin: new Thickness(8),
                    Shape: ShapeBorder.RoundedRectangle(18),
                    ClipBehavior: _clip ? Clip.AntiAlias : Clip.None)
                : new CardThemeData()
        };

        return new Theme(
            data: pageTheme,
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("Card baseline", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Elevated, filled, and outlined Material card variants with theme, mode, and clip probes.",
                        fontSize: 14,
                        color: Color.Parse("#8A000000")),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: _useMaterial3 ? "M3" : "M2",
                                onTap: () => SetState(() => _useMaterial3 = !_useMaterial3),
                                width: 80,
                                background: Color.Parse("#FFE9F0FF")),
                            BuildControlButton(
                                label: _useThemeOverrides ? "Theme on" : "Theme off",
                                onTap: () => SetState(() => _useThemeOverrides = !_useThemeOverrides),
                                width: 112,
                                background: Color.Parse("#FFEAF6F7")),
                            BuildControlButton(
                                label: _clip ? "Clip on" : "Clip off",
                                onTap: () => SetState(() => _clip = !_clip),
                                width: 96,
                                background: Color.Parse("#FFF0E8FF")),
                        ]),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: _dense ? "Dense" : "Regular",
                                onTap: () => SetState(() => _dense = !_dense),
                                width: 98,
                                background: Color.Parse("#FFF8EFE2")),
                            BuildControlButton(
                                label: "Reset",
                                onTap: ResetState,
                                width: 88,
                                background: Color.Parse("#FFF3E8D8")),
                        ]),
                    new Text(
                        $"useMaterial3={(_useMaterial3 ? "true" : "false")}, theme={(_useThemeOverrides ? "true" : "false")}, clip={(_clip ? "true" : "false")}, dense={(_dense ? "true" : "false")}",
                        fontSize: 12,
                        color: Color.Parse("#FF607D8B")),
                    new Expanded(
                        child: new Container(
                            color: Color.Parse("#FFF7F9FC"),
                            child: new SingleChildScrollView(
                                child: new Column(
                                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                                    children:
                                    [
                                        BuildElevatedCard(),
                                        BuildFilledCard(),
                                        BuildOutlinedCard(),
                                    ])))),
                ]));
    }

    private Widget BuildElevatedCard()
    {
        return new Card(
            clipBehavior: _clip ? Clip.AntiAlias : null,
            child: new ListTile(
                dense: _dense,
                leading: new Icon(Icons.StarOutline),
                title: new Text("Elevated card"),
                subtitle: new Text("Default variant keeps elevation and surfaceContainerLow color."),
                trailing: new Icon(Icons.InfoOutline)));
    }

    private Widget BuildFilledCard()
    {
        return Card.Filled(
            clipBehavior: _clip ? Clip.AntiAlias : null,
            child: BuildCardBody(
                title: "Filled card",
                body: "Filled cards use a quieter container color and zero default elevation in Material 3."));
    }

    private Widget BuildOutlinedCard()
    {
        return Card.Outlined(
            clipBehavior: _clip ? Clip.AntiAlias : null,
            child: BuildCardBody(
                title: "Outlined card",
                body: "Outlined cards add the default outlineVariant border while keeping elevation at zero."));
    }

    private Widget BuildCardBody(string title, string body)
    {
        return new Container(
            padding: _dense ? new Thickness(14, 10) : new Thickness(18, 14),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Start,
                spacing: 6,
                children:
                [
                    new Text(title, fontSize: 16, color: Colors.Black),
                    new Text(body, fontSize: 13, color: Color.Parse("#FF607D8B")),
                ]));
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                backgroundColor: background,
                foregroundColor: Colors.Black,
                minHeight: 36,
                padding: new Thickness(10, 8),
                borderRadius: BorderRadius.Circular(8),
                child: new Text(label, fontSize: 12)));
    }

    private void ResetState()
    {
        SetState(() =>
        {
            _useMaterial3 = true;
            _useThemeOverrides = false;
            _clip = false;
            _dense = false;
        });
    }
}

File: dart_sample/lib/demos/material/card_demo_page.dart
import 'package:flutter/material.dart';

class CardDemoPage extends StatefulWidget {
  const CardDemoPage({super.key});

  @override
  State<CardDemoPage> createState() => _CardDemoPageState();
}

class _CardDemoPageState extends State<CardDemoPage> {
  bool _useMaterial3 = true;
  bool _useThemeOverrides = false;
  bool _clip = false;
  bool _dense = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData baseTheme = Theme.of(context);
    final ThemeData pageTheme = ThemeData(
      useMaterial3: _useMaterial3,
      colorScheme: baseTheme.colorScheme,
      textTheme: baseTheme.textTheme,
      scaffoldBackgroundColor: baseTheme.scaffoldBackgroundColor,
      cardTheme: _useThemeOverrides
          ? CardThemeData(
              color: const Color(0xFFF5F9EE),
              shadowColor: const Color(0xFF455A64),
              surfaceTintColor: const Color(0xFF6750A4),
              elevation: 3,
              margin: const EdgeInsets.all(8),
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(18),
              ),
              clipBehavior: _clip ? Clip.antiAlias : Clip.none,
            )
          : const CardThemeData(),
    );

    return Theme(
      data: pageTheme,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 10,
        children: <Widget>[
          const Text(
            'Card baseline',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const Text(
            'Card variants plus Material surfaces and mergeable slices with theme, mode, and clip probes.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Row(
            spacing: 8,
            children: <Widget>[
              _buildControlButton(
                label: _useMaterial3 ? 'M3' : 'M2',
                onTap: () => setState(() => _useMaterial3 = !_useMaterial3),
                width: 80,
                background: const Color(0xFFE9F0FF),
              ),
              _buildControlButton(
                label: _useThemeOverrides ? 'Theme on' : 'Theme off',
                onTap: () =>
                    setState(() => _useThemeOverrides = !_useThemeOverrides),
                width: 112,
                background: const Color(0xFFEAF6F7),
              ),
              _buildControlButton(
                label: _clip ? 'Clip on' : 'Clip off',
                onTap: () => setState(() => _clip = !_clip),
                width: 96,
                background: const Color(0xFFF0E8FF),
              ),
            ],
          ),
          Row(
            spacing: 8,
            children: <Widget>[
              _buildControlButton(
                label: _dense ? 'Dense' : 'Regular',
                onTap: () => setState(() => _dense = !_dense),
                width: 98,
                background: const Color(0xFFF8EFE2),
              ),
              _buildControlButton(
                label: 'Reset',
                onTap: _resetState,
                width: 88,
                background: const Color(0xFFF3E8D8),
              ),
            ],
          ),
          Text(
            'useMaterial3=$_useMaterial3, theme=$_useThemeOverrides, clip=$_clip, dense=$_dense',
            style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
          ),
          Expanded(
            child: ColoredBox(
              color: const Color(0xFFF7F9FC),
              child: SingleChildScrollView(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: <Widget>[
                    _buildElevatedCard(),
                    _buildFilledCard(),
                    _buildOutlinedCard(),
                    _buildMaterialSurface(),
                    _buildMergeableMaterial(),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildElevatedCard() {
    return Card(
      clipBehavior: _clip ? Clip.antiAlias : null,
      child: ListTile(
        dense: _dense,
        leading: const Icon(Icons.star_outline),
        title: const Text('Elevated card'),
        subtitle: const Text(
          'Default variant keeps elevation and surfaceContainerLow color.',
        ),
        trailing: const Icon(Icons.info_outline),
      ),
    );
  }

  Widget _buildFilledCard() {
    return Card.filled(
      clipBehavior: _clip ? Clip.antiAlias : null,
      child: _buildCardBody(
        title: 'Filled card',
        body:
            'Filled cards use a quieter container color and zero default elevation in Material 3.',
      ),
    );
  }

  Widget _buildOutlinedCard() {
    return Card.outlined(
      clipBehavior: _clip ? Clip.antiAlias : null,
      child: _buildCardBody(
        title: 'Outlined card',
        body:
            'Outlined cards add the default outlineVariant border while keeping elevation at zero.',
      ),
    );
  }

  Widget _buildMaterialSurface() {
    return Container(
      margin: const EdgeInsets.all(4),
      child: Material(
        type: MaterialType.card,
        color: const Color(0xFFFFF8E1),
        elevation: _useThemeOverrides ? 5 : 2,
        surfaceTintColor: _useMaterial3
            ? const Color(0xFF6750A4)
            : Colors.transparent,
        clipBehavior: _clip ? Clip.antiAlias : Clip.none,
        child: _buildCardBody(
          title: 'Material surface',
          body:
              'Card-type Material resolves surface tint, elevation, clipping, and its default text style.',
        ),
      ),
    );
  }

  Widget _buildMergeableMaterial() {
    return Container(
      margin: const EdgeInsets.all(4),
      child: MergeableMaterial(
        elevation: _useThemeOverrides ? 4 : 2,
        hasDividers: true,
        children: <MergeableMaterialItem>[
          MaterialSlice(
            key: const ValueKey<String>('material-slice-first'),
            color: const Color(0xFFE8F5E9),
            child: _buildCardBody(
              title: 'First mergeable slice',
              body: 'Slices join on a shared Material card surface.',
            ),
          ),
          MaterialGap(
            key: const ValueKey<String>('material-slice-gap'),
            size: _dense ? 8 : 16,
          ),
          MaterialSlice(
            key: const ValueKey<String>('material-slice-second'),
            color: const Color(0xFFE3F2FD),
            child: _buildCardBody(
              title: 'Second mergeable slice',
              body:
                  'The gap is controller-animated when the item list changes.',
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildCardBody({required String title, required String body}) {
    return Padding(
      padding: _dense
          ? const EdgeInsets.symmetric(horizontal: 14, vertical: 10)
          : const EdgeInsets.symmetric(horizontal: 18, vertical: 14),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        spacing: 6,
        children: <Widget>[
          Text(
            title,
            style: const TextStyle(fontSize: 16, color: Colors.black),
          ),
          Text(
            body,
            style: const TextStyle(fontSize: 13, color: Colors.blueGrey),
          ),
        ],
      ),
    );
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }

  void _resetState() {
    setState(() {
      _useMaterial3 = true;
      _useThemeOverrides = false;
      _clip = false;
      _dense = false;
    });
  }
}
